### PR TITLE
Add option for some crude tracing on `stdout`

### PR DESCRIPTION
**Pull request by larseggert**
## Description

This adds some code to trace operation on `stdout`. Mostly useful during development on platforms where no other tracing is supported (e.g., MacOS).

This also casts values to the types indicated in the format strings, because compilers otherwise warn (because of a mismatch between the format string and the variable type.)

## Testing

No new tests needed.

## Documentation

Nope?

### CODE REVIEW

File: src/core/datagram.c [hunk 1]
```diff
@@ -372,7 +372,7 @@ QuicDatagramQueueSend(
                 AllocFailure,
                 "Allocation of '%s' failed. (%llu bytes)",
                 "DATAGRAM_SEND operation",
-                0);
+                UINT64_C(0));
```
Comment (nibanks) [inline src/core/datagram.c:375]: Perhaps instead of all these, we use `0ULL`? That works, right?

**Comment by nibanks** [comment]
Is the build going to validate all these things now? If not, how do we prevent regressions from creeping in?

File: src/core/datagram.c (hunk 1 above)
Comment (larseggert) [reply src/core/datagram.c:375]: Yes, that would work. Will make the change.

**Comment by larseggert** [comment]
In order to validate it, we would need to add a build that uses `QUIC_EVENTS_STDOUT` and `QUIC_LOGS_STDOUT`. Do you want me to add stuff to `CMakeLists.txt` to make that selectable?

**Comment by larseggert** [comment]
Other idea: Are any of the current CI runs building with `QUIC_EVENTS_STUB` and `QUIC_LOGS_STUB`? Because then I cold just add ` __attribute__((format(printf, 1, 2)))` to the definition of the stub function to make the compilers check this.

**Comment by nibanks** [comment]
Yeah, probably worth exposing. We also have a number of builds under the `build_windows_nontest` and `build_linux_nontest` stages that you could leverage to validate this. Or just add a new one.

File: src/generated/stdout/quic_trace.c [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                     _Out_writes_bytes_(2 * BufferLen) char * HexString);
+
+
+char * casted_clog_bytearray(const uint8_t * const data,
+                             const size_t len,
+                             struct clog_param ** head)
+{
+    struct clog_param * const param =
+        CXPLAT_ALLOC_PAGED(sizeof(*param), QUIC_POOL_TMP_ALLOC);
+    CXPLAT_FRE_ASSERT(param);
```
Comment (nibanks) [inline src/generated/stdout/quic_trace.c:33]: You can't assert that allocations succeed.

File: src/generated/stdout/quic_trace.c [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    // return the string representation
+    return param->str;
+}
+
+
+void clog_stdout(struct clog_param * head, const char * const format, ...)
+{
+    // replace clog specifiers in the format string
+    const char * repls[] = {"!CID!", "!ADDR!", "!VNL!", "!ALPN!"};
+    char * reformat = strdup(format);
```
Comment (nibanks) [inline src/generated/stdout/quic_trace.c:67]: Can fail to allocate

File: src/core/inline.c [hunk 4]
```diff
@@ -800,3 +800,4 @@ QuicMtuDiscoveryCheckSearchCompleteTimeout(
     _In_ QUIC_CONNECTION* Connection,
     _In_ uint64_t TimeNow
     );
+
```
Comment (nibanks) [inline src/core/inline.c:803]: nit: unnecessary change?

File: CMakeLists.txt [hunk 5]
```diff
@@ -282,9 +286,14 @@ else()
...
         check_function_exists(sysctl HAS_SYSCTL)
+        if (QUIC_ENABLE_LOGGING AND QUIC_ENABLE_LOGGING_STDOUT EQUAL 0)
+            set (QUIC_ENABLE_LOGGING OFF)
+            message(STATUS "Disabling all tracing on MacOS")
+        endif()
     endif()
 
-    set(QUIC_COMMON_FLAGS "")
+    # for reproducible builds: https://reproducible-builds.org/docs/build-path/
+    set(QUIC_COMMON_FLAGS "-ffile-prefix-map=${CMAKE_SOURCE_DIR}=.")
```
Comment (nibanks) [inline CMakeLists.txt:296]: Is this necessary?

File: CMakeLists.txt [hunk 6]
```diff
@@ -104,6 +104,10 @@ option(QUIC_DISABLE_POSIX_GSO "Disable GSO for systems that say they support it
 option(QUIC_OFFICIAL_RELEASE "Configured the build for an official release" OFF)
 set(QUIC_FOLDER_PREFIX "" CACHE STRING "Optional prefix for source group folders when using an IDE generator")
 set(QUIC_LIBRARY_NAME "msquic" CACHE STRING "Override the output library name")
+set(QUIC_ENABLE_LOGGING_STDOUT 0 CACHE STRING "Enables logging on stdout at given verbosity (0=off, 4=debug)")
```
Comment (nibanks) [inline CMakeLists.txt:107]: Why go this route, and not just on/off? And later, we could add a `LOGGING_LEVEL` param if we want?

File: CMakeLists.txt [hunk 7]
```diff
@@ -567,16 +576,17 @@ else() #!WIN32
...
+    if (QUIC_ENABLE_SANITIZERS AND QUIC_ENABLE_LOGGING_STDOUT EQUAL 0)
         set(QUIC_ENABLE_LOGGING OFF)
         message(WARNING "LTTng logging is incompatible with sanitizers. Skipping logging")
     endif()
 
     if(QUIC_ENABLE_LOGGING)
-        if (APPLE)
-            set (QUIC_ENABLE_LOGGING OFF)
-            message(STATUS "Disabling all tracing on macos")
-            list(APPEND QUIC_COMMON_DEFINES QUIC_EVENTS_STUB QUIC_LOGS_STUB)
```
Comment (nibanks) [inline CMakeLists.txt:579]: Shouldn't we still be using stub if not stdout?

File: CMakeLists.txt (hunk 5 above)
Comment (larseggert) [reply CMakeLists.txt:296]: No. I'll do a different PR for it.

File: src/core/inline.c (hunk 4 above)
Comment (larseggert) [reply src/core/inline.c:803]: Yep, good catch

File: CMakeLists.txt (hunk 6 above)
Comment (larseggert) [reply CMakeLists.txt:107]: I revert to on/off if you prefer.

File: CMakeLists.txt (hunk 7 above)
Comment (larseggert) [reply CMakeLists.txt:579]: The "Disabling all tracing on MacOS" stuff around line 288 sets `QUIC_ENABLE_LOGGING` to `OFF`, which means the stubs are being used.

**Comment by larseggert** [comment]
One question on the stubs: the way they're currently defined (as inline functions) makes them show up in the compiled binary even when compiling for releases. Is this really what you want?

(I cold make the stubs an empty define, which would generate no code at all in that case.)

File: CMakeLists.txt [hunk 8]
```diff
@@ -263,28 +267,23 @@ else()
...
+                message(WARNING "Must explicitly set QUIC_LOGGING_TYPE to \"stdout\" to enable due to performance overhead; disabling logging")
                 set(QUIC_ENABLE_LOGGING OFF)
             endif()
         endif()
-
-    elseif(CX_PLATFORM STREQUAL "darwin")
-        check_function_exists(sysctl HAS_SYSCTL)
     endif()
 
-    set(QUIC_COMMON_FLAGS "")
```
Comment (nibanks) [inline CMakeLists.txt:287]: Why remove this line?

File: scripts/build.ps1 [hunk 9]
```diff
@@ -24,6 +24,9 @@ This script provides helpers for building msquic.
 .PARAMETER DisableLogs
     Disables log collection.
 
+.PARAMETER StdoutLogs
+    Log to stdout (Only supported on unix)
+
```
Comment (nibanks) [inline scripts/build.ps1:29]: We should make this match CMake options. Let's update this to set the logging type. It should be optional and default to an empty string. Only if it's set to a non-empty string do we pass it down to CMake.

File: scripts/build.ps1 [hunk 10]
```diff
@@ -405,8 +405,8 @@ function CMake-Generate {
     if (!$DisableLogs) {
         $Arguments += " -DQUIC_ENABLE_LOGGING=on"
     }
-    if ($StdoutLogs) {
-        $Arguments += " -DQUIC_ENABLE_LOGGING=on -DQUIC_LOGGING_TYPE=stdout"
+    if ($LoggingType) {
+        $Arguments += " -DQUIC_ENABLE_LOGGING=on -DQUIC_LOGGING_TYPE=" + $LoggingType
```
Comment (nibanks) [inline scripts/build.ps1:409]: Let's put the `$LoggingType` check in the `if (!$DisableLogs)` section. Then you don't need to set `QUIC_ENABLE_LOGGING` just for the type, which makes things confusing if somehow `DisableLogs` is set too.

File: .azure/azure-pipelines.ci.yml [hunk 11]
```diff
@@ -456,7 +456,7 @@ stages:
       arch: arm64
       tls: openssl
       extraName: 'stdout'
-      extraBuildArgs: -StdoutLogs
+      extraBuildArgs: -LoggingType=stdout
```
Comment (nibanks) [inline .azure/azure-pipelines.ci.yml:459]:
```suggestion
      extraBuildArgs: -LoggingType stdout
```

**Comment by nibanks** [comment]
Looks like you accidentally updated the googletest submodule.

File: scripts/build.ps1 [hunk 12]
```diff
@@ -399,6 +405,9 @@ function CMake-Generate {
     if (!$DisableLogs) {
         $Arguments += " -DQUIC_ENABLE_LOGGING=on"
     }
+    if ($LoggingType) {
```
Comment (nibanks) [inline scripts/build.ps1:408]:
```suggestion
    if ($LoggingType -ne "") {
```

**Review by nibanks — APPROVED** [review]
